vraptor-core/src/main/java/br/com/caelum/vraptor/observer/ExecuteMethod.java

```diff
@@ -79,6 +85,11 @@ public Void call() throws Exception {
 
 				log.debug("Invoking {}", reflectionMethod);
 				Object instance = event.getControllerInstance();
+				
+				Class<?> declaringClass = reflectionMethod.getDeclaringClass();
+				if (!declaringClass.isAssignableFrom(instance.getClass())) {
+					instance = container.instanceFor(declaringClass);
```

Comment:
I've looked defaultpageresult implementation and it looks like it does a
regular forward using request dispatcher...

Em 14:40 sex, 15/05/2015, Rodrigo Turini <notifications@github.com>
escreveu:

> In
> vraptor-core/src/main/java/br/com/caelum/vraptor/observer/ExecuteMethod.java
> <https://github.com/caelum/vraptor4/pull/978#discussion_r30429062>:
>
> > @@ -79,6 +85,11 @@ public Void call() throws Exception {
> >
> >  				log.debug("Invoking {}", reflectionMethod);
> >  				Object instance = event.getControllerInstance();
> > +				
> > +				Class<?> declaringClass = reflectionMethod.getDeclaringClass();
> > +				if (!declaringClass.isAssignableFrom(instance.getClass())) {
> > +					instance = container.instanceFor(declaringClass);
>
> the problem is that AFAIK we dont do a real forward, we just execute the
> new controller
> method... so in the ExecuteMethod the controllerInstance still uses the
> first instance.
> The weird think is that it works when we forward programmatically, just
> like this:
>
> result.forwardTo(this).other();
>
> but in that case, Mirror works automagically, even with the wrong class :)
> bug or behavior?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/vraptor4/pull/978/files#r30429062>.
>